analyzer/os/amazonlinux/amazonlinux.go

```diff
@@ -4,51 +4,63 @@ import (
 	"bufio"
 	"bytes"
 	"fmt"
+	"os"
 	"strings"
 
-	"github.com/aquasecurity/fanal/types"
-
 	"golang.org/x/xerrors"
 
-	"github.com/aquasecurity/fanal/analyzer/os"
+	"github.com/aquasecurity/fanal/utils"
+
+	"github.com/aquasecurity/fanal/types"
+
+	aos "github.com/aquasecurity/fanal/analyzer/os"
 
 	"github.com/aquasecurity/fanal/analyzer"
-	"github.com/aquasecurity/fanal/extractor"
 )
 
 func init() {
-	analyzer.RegisterOSAnalyzer(&amazonlinuxOSAnalyzer{})
+	analyzer.RegisterAnalyzer(&amazonlinuxOSAnalyzer{})
 }
 
+var requiredFiles = []string{"etc/system-release"}
+
 type amazonlinuxOSAnalyzer struct{}
 
-func (a amazonlinuxOSAnalyzer) Analyze(fileMap extractor.FileMap) (types.OS, error) {
-	for _, filename := range a.RequiredFiles() {
-		file, ok := fileMap[filename]
-		if !ok {
-			continue
-		}
-		scanner := bufio.NewScanner(bytes.NewBuffer(file))
-		for scanner.Scan() {
-			line := scanner.Text()
-			fields := strings.Fields(line)
-			// Only Amazon Linux Prefix
-			if strings.HasPrefix(line, "Amazon Linux release 2") {
-				return types.OS{
-					Family: os.Amazon,
-					Name:   fmt.Sprintf("%s %s", fields[3], fields[4]),
-				}, nil
-			} else if strings.HasPrefix(line, "Amazon Linux") {
-				return types.OS{
-					Family: os.Amazon,
-					Name:   strings.Join(fields[2:], " "),
-				}, nil
-			}
+func (a amazonlinuxOSAnalyzer) Analyze(content []byte) (analyzer.AnalyzeReturn, error) {
+	foundOS, err := a.parseRelease(content)
+	if err != nil {
+		return analyzer.AnalyzeReturn{}, err
+	}
+	return analyzer.AnalyzeReturn{
+		OS: foundOS,
+	}, nil
+}
+
+func (a amazonlinuxOSAnalyzer) parseRelease(content []byte) (types.OS, error) {
+	scanner := bufio.NewScanner(bytes.NewBuffer(content))
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		// Only Amazon Linux Prefix
+		if strings.HasPrefix(line, "Amazon Linux release 2") {
+			return types.OS{
+				Family: aos.Amazon,
+				Name:   fmt.Sprintf("%s %s", fields[3], fields[4]),
+			}, nil
+		} else if strings.HasPrefix(line, "Amazon Linux") {
+			return types.OS{
+				Family: aos.Amazon,
+				Name:   strings.Join(fields[2:], " "),
```

Comment:
We should check what size `fields` is before we operate on individual slice items.

Reply:
I've fixed it in https://github.com/aquasecurity/fanal/pull/110/commits/d1bdfd8fbc15a7c2333f86502e86161597fcee25
https://play.golang.org/p/K2DC8_7ajCW